gquant/dataframe_flow/task.py

```diff
@@ -0,0 +1,394 @@
+
+
+class TaskSpecSchema(object):
+    '''Outline fields expected in a dictionary specifying a task node.
+
+    :ivar id: unique id or name for the node
+    :ivar type: Plugin class i.e. subclass of Node. Specified as string
+        or subclass of Node
+    :ivar conf: Configuration for the plugin i.e. parameterization. This is a
+        dictionary.
+    :ivar filepath: Path to python module for custom plugin types.
+    :ivar inputs: List of ids of other tasks or an empty list.
+    '''
+
+    id = 'id'
+    type = 'type'
+    conf = 'conf'
+    filepath = 'filepath'
+    inputs = 'inputs'
+    load = 'load'
+    save = 'save'
+
+    @classmethod
+    def _typecheck(cls, schema_field, value):
+        if (schema_field == cls.id):
+            assert isinstance(value, str)
+        elif schema_field == cls.type:
+            assert (isinstance(value, str) or issubclass(value, Node))
+        elif schema_field == cls.conf:
+            assert (isinstance(value, dict) or isinstance(value, list))
+        elif schema_field == cls.filepath:
+            assert isinstance(value, str)
+        elif schema_field == cls.inputs:
+            assert isinstance(value, list)
+            for item in value:
+                assert isinstance(item, str)
+        elif schema_field == cls.load:
+            pass
+        elif schema_field == cls.save:
+            assert isinstance(value, bool)
+        else:
+            raise KeyError
+
+    _schema_req_fields = [id, type, conf, inputs]
+
+    @classmethod
+    def validate(cls, task_spec):
+        '''
+        :param task_spec: A dictionary per TaskSpecSchema
+        '''
+        for ifield in cls._schema_req_fields:
+            if ifield not in task_spec:
+                raise KeyError('task spec missing required field: {}'
+                               .format(ifield))
+
+        for task_field, field_val in task_spec.items():
+            cls._typecheck(task_field, field_val)
+
+
+class Task(object):
+    ''' A strong typed Task class that is converted from dictionary.
+    '''
+
+    def __init__(self, task_spec):
+        self._task_spec = {}  # internal dict
+        self._task_spec[TaskSpecSchema.id] = task_spec[TaskSpecSchema.id]
+        TaskSpecSchema.validate(task_spec)
+        self._task_spec[TaskSpecSchema.type] = task_spec[TaskSpecSchema.type]
+        self._task_spec[TaskSpecSchema.conf] = task_spec[TaskSpecSchema.conf]
+        if TaskSpecSchema.filepath in task_spec:
+            self._task_spec[TaskSpecSchema.filepath] = task_spec[
+                TaskSpecSchema.filepath]
+        if TaskSpecSchema.load in task_spec:
+            self._task_spec[TaskSpecSchema.load] =\
+                task_spec[TaskSpecSchema.load]
+        if TaskSpecSchema.save in task_spec:
+            self._task_spec[TaskSpecSchema.save] =\
+                task_spec[TaskSpecSchema.save]
+        self._task_spec[TaskSpecSchema.inputs] =\
+            task_spec[TaskSpecSchema.inputs]
```

Comment:
agree. I deep copied the spec file.

Reply:
I'm not sure if everything should be deepcopied. If I pass a cudf dataframe for the load, my intention typically is not to deepcopy it because that will use up more memory on the GPU.

Maybe we should add another spec field to the task:
```
class TaskSpecSchema(object):
    '''
    :cvar copy: Copy option when instantiating a task, either deep or shallow.
        Assumed shallow by default.
    '''
    copy = 'copy'

task_spec = {
    TaskSpecSchema.copy: 'deep'
}
```

Default would be shallow. That way the end-user is aware and can make their choice as appropriate. They would be responsible to not mess with mutable references if passing them in.

Reply:
Yes. deepcopy of dataframe can be a problem. We can deepcopy everything except the "load" field

Reply:
Typically in these situations just inform the end-user so they are aware. To keep it simple I would either deepcopy everything or shallow copy everything. The end-user can make their choice. They can always make their own deepcopies if needed. Most of the time shallow copy will be fine I think.

The code seems unnecessarily verbose.
https://github.com/yidong72/gQuant/blob/fcade400d5c70c6b29a607407d3c69b83ac8cd5a/gquant/dataframe_flow/task.py#L17-L37

If the task_spec is validated, just copy it. No need to check every single field. That make it difficult in the future is the API is extended.
```
    def __init__(self, task_spec):
        TaskSpecSchema.validate(task_spec)
        self._task_spec = deepcopy(task_spec)
        # or if making it optional
        # if task_spec.get(TaskSpecSchema.copy, 'shallow') == 'deep':
        #     self._task_spec = deepcopy(task_spec)
        # else:
        #     self._task_spec = task_spec
```

Reply:
deep copy of large GPU dataframe is very expensive, we want to avoid as much as possible. And I will try to make as many choices for the user as possible, as we are going to assume the users are data scientists who don't want to deal with memory management at all.

Reply:
Maybe just leave it as a shallow copy. If there is a side-effect bug, the end-user will discover it and make their own deepcopies as necessary. Can you simplify this? There's no need to copy every single field like that. The validate routine is in charge of checking for existence and valid types so it's redundant.
Instead of this:
```
    def __init__(self, task_spec):
        self._task_spec = {}  # internal dict
        TaskSpecSchema.validate(task_spec)
        # just deepcopy the inputs and conf
        self._task_spec[TaskSpecSchema.task_id] =\
            task_spec[TaskSpecSchema.task_id]
        self._task_spec[TaskSpecSchema.node_type] =\
            task_spec[TaskSpecSchema.node_type]
        self._task_spec[TaskSpecSchema.conf] =\
            copy.deepcopy(task_spec[TaskSpecSchema.conf])
        if TaskSpecSchema.filepath in task_spec:
            self._task_spec[TaskSpecSchema.filepath] = task_spec[
                TaskSpecSchema.filepath]
        if TaskSpecSchema.load in task_spec:
            self._task_spec[TaskSpecSchema.load] =\
                task_spec[TaskSpecSchema.load]
        if TaskSpecSchema.save in task_spec:
            self._task_spec[TaskSpecSchema.save] =\
                task_spec[TaskSpecSchema.save]
        self._task_spec[TaskSpecSchema.inputs] =\
            copy.deepcopy(task_spec[TaskSpecSchema.inputs])
```
Do something simpler:
```
    def __init__(self, task_spec):
        # whatever is passed in has to be valid
        TaskSpecSchema.validate(task_spec)
        self._task_spec = copy.copy(task_spec)
        # deepcopies of inputs can still be done
        self._task_spec[TaskSpecSchema.inputs] = \
            copy.deepcopy(task_spec[TaskSpecSchema.inputs])
        # imo, I don't think conf should always be deepcopied
        self._task_spec[TaskSpecSchema.conf] =\
            copy.deepcopy(task_spec[TaskSpecSchema.conf])
```

If you don't want to simplify this, can you explain why you want to copy every field like that?

Reply:
yes. we can use copy instead of setting each field. deepcopy of conf is light-weighted, it is safer this way.